File: module/service/middleware_sub_test_service/middleware_sub_test/node.py
import rclpy
import importlib
import sys
from rclpy.node import Node
from ament_index_python.packages import get_package_share_directory
from pathlib import Path
from pacific_rim_log import info
from pacific_rim_metric import counter, runtime_metric_names
from pacific_rim_otel import init_observability
from pacific_rim_trace import start_span
from pacific_rim_communication_infra import CommunicationRuntimeThread


def load_generated_register():
  for root in Path(__file__).resolve().parents:
    generated_root = root / "pkg" / "idl" / "middleware_sub_test_service" / "generated"
    candidate = generated_root / "python" / "register.py"
    if candidate.exists():
      sys.path.insert(0, str(generated_root))
      module = importlib.import_module("python.register")
      return module.register_generated_interfaces
  async def register_generated_interfaces(runtime, provider=None):
    _ = runtime
    _ = provider
  return register_generated_interfaces


class MiddlewareSubTestNode(Node):
  def __init__(self):
    super().__init__("middleware_sub_test")
    self.observability = init_observability("middleware_sub_test")
    self._communication_thread = None
    self.communication = self.start_communication()
    self.run_communication(load_generated_register()(self.communication))
    span = start_span("middleware_sub_test.startup")
    counter(runtime_metric_names["message_count"]).add()
    info(
      "Middleware Sub Test node started",
      {
        "traceId": span.trace_id,
        "spanId": span.span_id,
      },
    )
    span.end()
    # After adding pkg/idl contracts and config routes, run tools/generate-interfaces.sh.
    # Generated route bindings live in pkg/idl; middleware bootstrap stays in infra.

  def start_communication(self):
    config_path = Path(get_package_share_directory("middleware_sub_test")) / "config" / "config.yaml"
    self._communication_thread = CommunicationRuntimeThread(config_path)
    runtime = self._communication_thread.start()
    info("Middleware Sub Test communication started", {"config": str(config_path)})
    return runtime

  def run_communication(self, coroutine):
    if self._communication_thread is None:
      raise RuntimeError("communication runtime is not started")
    return self._communication_thread.run(coroutine)

  def stop_communication(self):
    if self._communication_thread is not None:
      self._communication_thread.stop()
      self._communication_thread = None
    self.communication = None


def main(args=None):
  rclpy.init(args=args)
  node = MiddlewareSubTestNode()

  try:
    rclpy.spin(node)
  except KeyboardInterrupt:
    pass
  finally:
    node.stop_communication()
    node.observability.shutdown()
    node.destroy_node()
    if rclpy.ok():
      rclpy.shutdown()


if __name__ == "__main__":
  main()

File: module/service/middleware_rpc_client_test_service/src/node.cpp
#include <memory>

#include "ament_index_cpp/get_package_share_directory.hpp"
#include "infra/communication/cpp/core/bootstrap.hpp"
#include "pacific_rim/log/log.hpp"
#include "pacific_rim/metric/metric.hpp"
#include "pacific_rim/trace/trace.hpp"
#include "rclcpp/rclcpp.hpp"
#include "runtime/ros2/generated_interface_registry.hpp"

class MiddlewareRpcClientTestNode : public rclcpp::Node {
public:
  MiddlewareRpcClientTestNode() : Node("middleware_rpc_client_test") {
    const auto config_path =
        ament_index_cpp::get_package_share_directory("middleware_rpc_client_test") +
        "/config/config.yaml";
    communication_runtime_ =
        std::make_unique<pacific_rim::communication::core::CommunicationRuntime>(
            pacific_rim::communication::core::BootstrapCommunication(
                config_path,
                "middleware_rpc_client_test"));
    generated_interfaces_ =
        pacific_rim::generated::middleware_rpc_client_test::register_generated_interfaces(
            *this,
            *communication_runtime_);

    auto span = pacific_rim::trace::start_span("middleware_rpc_client_test.startup");
    startup_counter_.add();
    pacific_rim::log::info(
        get_logger(),
        "Middleware Rpc Client Test node started",
        {
            {"traceId", span.ids().trace_id},
            {"spanId", span.ids().span_id},
        });
    span.end();
  }

private:
  // After adding pkg/idl contracts and routes, generate a registry with:
  // ./tools/generate-interfaces.sh
  // Generated route bindings live in this module. Shared protocol abstractions
  // live in pkg/idl. Keep business logic outside node.cpp.
  pacific_rim::metric::Counter startup_counter_{
      pacific_rim::metric::message_count};
  std::unique_ptr<pacific_rim::communication::core::CommunicationRuntime>
      communication_runtime_;
  pacific_rim::generated::middleware_rpc_client_test::GeneratedInterfaceHandles
      generated_interfaces_;
};

int main(int argc, char **argv) {
  pacific_rim::communication::dds::EnsureCycloneDdsRmw();
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<MiddlewareRpcClientTestNode>());
  rclcpp::shutdown();
  return 0;
}
